crates/gen-guest-teavm-java/src/lib.rs

```diff
@@ -0,0 +1,1918 @@
+            ),
+
+            Instruction::ResultLift { result, ty } => self.lift_variant(
+                &Type::Id(*ty),
+                &[("ok", result.ok), ("err", result.err)],
+                iface,
+                &operands[0],
+                results,
+            ),
+
+            Instruction::EnumLower { .. } => results.push(format!("{}.ordinal()", operands[0])),
+
+            Instruction::EnumLift { name, .. } => results.push(format!(
+                "{}.values()[{}]",
+                name.to_upper_camel_case(),
+                operands[0]
+            )),
+
+            Instruction::ListCanonLower { element, realloc } => {
+                let op = &operands[0];
+                let (size, ty) = list_element_info(element);
+
+                // Note that we can only reliably use `Address.ofData` for elements with alignment <= 4 because as
+                // of this writing TeaVM does not guarantee 64 bit items are aligned on 8 byte boundaries.
+                if realloc.is_none() && size <= 4 {
+                    results.push(format!("Address.ofData({op}).toInt()"));
+                } else {
+                    let address = self.locals.tmp("address");
+                    let ty = ty.to_upper_camel_case();
+
+                    uwrite!(
+                        self.src,
+                        "
+                        Address {address} = Memory.malloc({size} * ({op}).length, {size});
+                        Memory.put{ty}s({address}, {op}, 0, ({op}).length);
+                        "
+                    );
+
+                    if realloc.is_none() {
+                        self.cleanup.push(Cleanup {
+                            address: format!("{address}.toInt()"),
+                            size,
+                            align: size,
+                        });
+                    }
+
+                    results.push(format!("{address}.toInt()"));
+                }
+                results.push(format!("({op}).length"));
+            }
+
+            Instruction::ListCanonLift { element, .. } => {
+                let (_, ty) = list_element_info(element);
+                let ty_upper = ty.to_upper_camel_case();
+                let array = self.locals.tmp("array");
+                let address = &operands[0];
+                let length = &operands[1];
+
+                uwrite!(
+                    self.src,
+                    "
+                    {ty}[] {array} = new {ty}[{length}];
+                    Memory.get{ty_upper}s(Address.fromInt({address}), {array}, 0, ({array}).length);
+                    "
+                );
+
+                results.push(array);
+            }
+
+            Instruction::StringLower { realloc } => {
+                let op = &operands[0];
+                let bytes = self.locals.tmp("bytes");
+                uwriteln!(
+                    self.src,
+                    "byte[] {bytes} = ({op}).getBytes(StandardCharsets.UTF_8);"
```

Comment:
Are strings natively encoded as utf-8 in java?

Reply:
The JRE can technically encode them however it likes, and I believe the `String` implementation in OpenJDK has changed several times, but since `char` is 16-bits, the "most native" encoding is UTF-16, and TeaVM uses a `char[]` to implement `String`.

Reply:
Ah ok, I'm wondering if we can get a good use case for exercising the utf16 option of guests or similar, and it sounds like maybe? That's fine to happen at a later date though, this repo isn't prepared to actually do that yet.